### ISSUE TITLE
For model compression:

### ISSUE THREAD

**Issue by Luadoo** [roymiles/ITRD#1]
Hello, sir

I want to know your paper experiment part 5, there has a description of quantization models like ReCU, IR-Net, and RBNN with KD. But in your code, there doesn't have any relation to source coding with the binary quantization model. Can you give more detail about how it could work? Thank you, so much!

**Comment by roymiles** [comment]
Hey!

We used the ReCU provided code (https://github.com/z-hXu/ReCU) and simply added in our distillation loss between the pre-trained frozen full precision model and the binary ReCU model being trained.

You can see in our helper/loops.py how the training pipeline is setup for the teacher/student setting. In general it is just instantiating the ITLoss class inside the student and then adding the losses during training between the teacher and the student representations https://github.com/roymiles/ITRD/blob/8e264eb8cce0b5d53a3ab531bdf553b0e1da1ce6/distiller_zoo/IT.py#L49

We did not implement IR-Net or RBNN with KD. Implementing ReviewKD is a lot more involved as you need intermediate losses between both networks. 

If you have any more questions, feel free to ask!

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/69e92344-be92-464d-833e-11a626f103db)
But, I don't know how to train this part experiment, can you give me more guidance? I have checked your source code for every module, but I didn't find the way, please give me more information is good for me. Thanks a lot! My email: jiexu2017action@gmail.com

**Comment by Luadoo** [comment]
Can you provide resnet18 in cifar10 pre-training weitgh PTH. file? Because the binary student you are just using is from the ReCU module, so can you share the full coding of how it works? Thank you, so much!

**Comment by roymiles** [comment]
Ok sure I have cleaned up the code and shared it via google drive (link below). I will say that I have not tested this in a very long time, but it should be a good starting point if you would like to look into this further!

https://drive.google.com/file/d/1WJ_rGIsQ-SaqvXzNsfkxcNA6rWBbcnD8/view?usp=sharing

**Comment by Luadoo** [comment]
Yes, thank you, sir! I will depend on your previous work to do more things! After some time, maybe have another question for you may I will request it again? If it doesn't spend your much time!

**Comment by Luadoo** [comment]
Yes, thank you, sir! When I try using the binary module get some error, I am not sure about the ReCU part, it was just like directly using the models_cifar file right?
![image](https://github.com/roymiles/ITRD/assets/58927660/2008a8bc-bf29-451a-a989-901969b3c93c)
Just using ReCU code like this is right?
![image](https://github.com/roymiles/ITRD/assets/58927660/9d60d8f7-0a26-45ea-b16c-dee694c61842)

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/6fe8e8fe-e3da-43f8-af19-5e6d9744b8c6)
Thank you for you provide the coding, but I get some errors, I am not sure how to train the paper method ReCU+L(corr)+L(mi),
![image](https://github.com/roymiles/ITRD/assets/58927660/a6123d08-5cf8-48ca-a3d2-ff01e5c73cc9)
So if the same paper gets 94.1% accu in cifar10, which KD methods should be imported? And from the "models_cifar" module I just used ReCU source coding it was right?

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/63853c47-8d70-4927-a662-ebb5c5722076)
Thanks again, I have just imported the 'models_cifar' module from ReCU coding, and when I try running the binary student gets some error, I don't find 'crd' model and also try using the 'IT' model for distillation but it was doesn't work. Would you have time to guide me to do better? Sir

**Comment by roymiles** [comment]
I am really sorry for the delay! I have been a bit caught up with stuff. It seems I forgot to add these folders to the zip. Here is the new link:
https://drive.google.com/file/d/1WJ_rGIsQ-SaqvXzNsfkxcNA6rWBbcnD8/view?usp=sharing

It seems to run correctly now, though I have no tested the results yet. I might leave a run going to see later today.

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/1fdf24f8-dc91-4873-9e3f-a6a4f2559dfb)
Yes, if you can rerun it to check is better, and maybe you forget to upload the Res18, Res20, and Vgg teacher pretraining weight pth. file. Thank you, so much!

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/edd74cf7-3be2-42b1-b5a4-e1f8b32cd1c7)
Sorry, sir, I don't find the Res18 pretraining weight pth. file, can you provide it? If you are perfectly working can you check it again? Thanks!

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/0406e928-0348-4e93-9068-ff1a91cecedd)

![image](https://github.com/roymiles/ITRD/assets/58927660/32971b62-6871-48ff-a82b-23a9f2abb1f7)
I have changed the resnet18.pth rooting, but it shows another error again, I am not sure what's the problem, can you give me some help? Thank you!

**Comment by roymiles** [comment]
It seems this code was using an old teacher checkpoint that I no longer have available. Instead of re-training a teacher I have updated the code to automatically download a pre-trained resnet18 from hugging face (timm version 0.6.12).
However, I will say that now the code will now only work with the ITRD loss. To get it working with the other distillation methods would require adding in additional hooks to extract the intermediate teacher features.

https://drive.google.com/file/d/1WJ_rGIsQ-SaqvXzNsfkxcNA6rWBbcnD8/view?usp=sharing

I am using the following command seems to train the student:
` python -u main.py --gpus 0 --results_dir ./result --data_path [DATA_PATH] --dataset cifar10 --epochs 600 --lr 0.1 -b 256 -bt 128 --lr_type cos --warm_up --weight_decay 5e-4 --tau_min 0.85 --tau_max 0.99 --save 'relu_resnet18' --lambda_corr 2 --lambda_mi 1 --kd_type 'itrd' --model 'resnet18_1w1a'`

and I get the following output in my terminal:

Evaluate teacher
```
EVALUATING - Epoch: [0][0/79]   Time 0.306 (0.306)      Data 0.151 (0.151)      Loss 0.2052 (0.2052)    Prec@1 95.312 (95.312)  Prec@5 100.000 (100.000)
Teacher acc: 94.980
lr: 0.02
TRAINING - Epoch: [0][0/196]    Time 0.847 (0.847)      Data 0.118 (0.118)      Loss 2.4326 (2.4326)    Prec@1 10.156 (10.156)  Prec@5 49.609 (49.609)
TRAINING - Epoch: [0][100/196]  Time 0.100 (0.108)      Data 0.000 (0.001)      Loss 1.9522 (2.2485)    Prec@1 32.812 (24.973)  Prec@5 81.641 (75.507)
EVALUATING - Epoch: [0][0/79]   Time 0.096 (0.096)      Data 0.083 (0.083)      Loss 1.8145 (1.8145)    Prec@1 33.594 (33.594)  Prec@5 86.719 (86.719)
Time cost: 00:21        Time of Finish: 2023-08-01 02:26:33

 Epoch: 1       Training Loss 2.0805    Training Prec@1 29.838  Training Prec@5 80.578  Validation Loss 1.9639  Validation Prec@1 33.420 Validation Prec@5 84.240
Best Prec@1 33.420

lr: 0.04
TRAINING - Epoch: [1][0/196]    Time 0.200 (0.200)      Data 0.140 (0.140)      Loss 1.9361 (1.9361)    Prec@1 35.156 (35.156)  Prec@5 85.156 (85.156)
TRAINING - Epoch: [1][100/196]  Time 0.099 (0.102)      Data 0.000 (0.002)      Loss 1.5962 (1.8873)    Prec@1 44.531 (37.051)  Prec@5 92.969 (87.573)
EVALUATING - Epoch: [1][0/79]   Time 0.110 (0.110)      Data 0.098 (0.098)      Loss 1.6087 (1.6087)    Prec@1 46.094 (46.094)  Prec@5 87.500 (87.500)
Time cost: 00:21        Time of Finish: 2023-08-01 02:17:34

 Epoch: 2       Training Loss 1.7696    Training Prec@1 40.658  Training Prec@5 89.146  Validation Loss 1.6019  Validation Prec@1 44.510 Validation Prec@5 91.400
Best Prec@1 44.510

lr: 0.06000000000000001
TRAINING - Epoch: [2][0/196]    Time 0.195 (0.195)      Data 0.146 (0.146)      Loss 1.5494 (1.5494)    Prec@1 47.656 (47.656)  Prec@5 91.797 (91.797)
TRAINING - Epoch: [2][100/196]  Time 0.102 (0.101)      Data 0.000 (0.002)      Loss 1.4023 (1.5978)    Prec@1 50.391 (47.006)  Prec@5 95.703 (92.068)
EVALUATING - Epoch: [2][0/79]   Time 0.097 (0.097)      Data 0.084 (0.084)      Loss 1.3759 (1.3759)    Prec@1 55.469 (55.469)  Prec@5 93.750 (93.750)
Time cost: 00:21        Time of Finish: 2023-08-01 02:18:05

 Epoch: 3       Training Loss 1.5102    Training Prec@1 50.080  Training Prec@5 92.830  Validation Loss 1.3912  Validation Prec@1 53.620 Validation Prec@5 93.780
Best Prec@1 53.620
...
```

I am leaving it running overnight and will update this comment with the results attained. 
Achieved 93.83 acc v.s. 93.90 in the paper using Lcorr only.

**Comment by Luadoo** [comment]
yeah, now is working perfectly! Thanks for your time. I will try it！

**Comment by Luadoo** [comment]
![image](https://github.com/roymiles/ITRD/assets/58927660/f44b7087-56d7-40ae-96cc-6ce52c678fcb)
Hello, sir I have run 3 times ITRD-ResNet18 but didn't get your paper showing the accuracy of 94.1%, so do you have any guidance for me?
